src/audio/google/google_ctc_audio_processing.c

```diff
@@ -0,0 +1,574 @@
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	uint8_t *config;
+	size_t size;
+	int ret;
+
+	comp_dbg(dev, "google_ctc_audio_processing_reconfigure()");
+
+	config = comp_get_data_blob(cd->tuning_handler, &size, NULL);
+	if (size == 0) {
+		/* No data to be handled */
+		return 0;
+	}
+
+	if (!config) {
+		comp_err(dev, "google_ctc_audio_processing_reconfigure(): Tuning config not set");
+		return -EINVAL;
+	}
+
+	comp_info(dev, "google_ctc_audio_processing_reconfigure(): New tuning config %p (%zu bytes)",
+		  config, size);
+
+	cd->reconfigure = false;
+
+	uint8_t *processing_config;
+	size_t processing_config_size;
+	bool processing_config_present;
+
+	GoogleCtcAudioProcessingParseSofConfigMessage(config, size,
+						      &processing_config,
+						      &processing_config_size,
+						      &processing_config_present);
+
+	if (processing_config_present) {
+		comp_info(dev,
+			  "google_ctc_audio_processing_reconfigure(): Applying config of size %zu bytes",
+			  processing_config_size);
+
+		ret = GoogleCtcAudioProcessingReconfigure(cd->state,
+							  processing_config,
+							  processing_config_size);
+		if (ret) {
+			comp_err(dev, "GoogleCtcAudioProcessingReconfigure failed: %d",
+				 ret);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int ctc_prepare(struct processing_module *mod,
+		       struct sof_source **sources, int num_of_sources,
+		       struct sof_sink **sinks, int num_of_sinks)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	struct comp_buffer *source;
+	int ret;
+
+	comp_info(mod->dev, "ctc_prepare()");
+
+	source = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+	cd->ctc_func = ctc_find_func(audio_stream_get_frm_fmt(&source->stream));
+	if (!cd->ctc_func) {
+		comp_err(mod->dev, "ctc_prepare(): No suitable processing function found.");
+		ret = -EINVAL;
+		goto err;
+	}
+
+	return 0;
+err:
+	return ret;
+}
+
+static int ctc_reset(struct processing_module *mod)
+{
+	comp_dbg(mod->dev, "ctc_reset()");
+
+	return 0;
+}
+
+static int ctc_process(struct processing_module *mod,
+		       struct input_stream_buffer *input_buffers,
+		       int num_input_buffers,
+		       struct output_stream_buffer *output_buffers,
+		       int num_output_buffers)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+	struct audio_stream *source = input_buffers[0].data;
+	struct audio_stream *sink = output_buffers[0].data;
+	uint32_t frames = input_buffers[0].size;
+
+	int ret;
+
+	comp_dbg(mod->dev, "ctc_process()");
+
+	if (cd->reconfigure) {
+		ret = google_ctc_audio_processing_reconfigure(mod);
+		if (ret)
+			return ret;
+	}
+
+	cd->ctc_func(cd, source, sink, &input_buffers[0], &output_buffers[0], frames);
+	return 0;
+}
+
+#if CONFIG_IPC_MAJOR_3
```

Comment:
Recent SOF idiom for this kind of thing, where you have large APIs that differ between IPC3/4, is to put the divergent handler code in separate files with a unified API to avoid the big block ifdefs.

Reply:
Split to google_ctc_audio_processing_ipc{3, 4}.c